drv/local-vpd/src/lib.rs

```diff
@@ -142,35 +158,33 @@ pub fn read_config_into(
                 if chunk.header().tag == *b"FRU0" {
                     chunk
                         .check_body_checksum(&mut scratch)
-                        .map_err(|_| err(LocalVpdError::InvalidChecksum))?;
+                        .map_err(LocalVpdError::InvalidChecksum)?;
                     let mut inner = chunk.read_as_chunks();
                     while let Ok(Some(chunk)) = inner.next() {
                         if chunk.header().tag == tag {
                             chunk.check_body_checksum(&mut scratch).map_err(
-                                |_| err(LocalVpdError::InvalidChecksum),
+                                LocalVpdError::InvalidChecksum,
                             )?;
 
                             let chunk_len = chunk.len() as usize;
 
                             if chunk_len > out.len() {
-                                return Err(err(
-                                    LocalVpdError::InvalidChunkSize,
-                                ));
+                                return Err(LocalVpdError::InvalidChunkSize);
                             }
 
                             chunk
                                 .read_exact(0, &mut out[..chunk_len])
-                                .map_err(|_| LocalVpdError::DeviceError)?;
+                                .map_err(LocalVpdError::ErrorOnRead)?;
                             return Ok(chunk_len);
                         }
                     }
-                    return Err(err(LocalVpdError::NoSuchChunk));
+                    return Err(LocalVpdError::NoSuchChunk(tag));
                 } else {
                     ringbuf_entry!(Trace::UnrelatedChunk(chunk.header().tag));
                 }
             }
-            Ok(None) => return Err(err(LocalVpdError::NoRootChunk)),
-            Err(_) => return Err(err(LocalVpdError::BadRootChunk)),
+            Ok(None) => return Err(LocalVpdError::NoRootChunk),
```

Comment:
I don't remember exactly how `reader.next()` behaves, but it _looks_ like this error might be wrong. We could have a root chunk, go into the `Ok(Some(chunk))` branch some number of times, and then reach the end and hit this branch, right? Unless `reader.next()` is guaranteed to return an error instead of `Ok(None)` when it gets to the end (which might be how it actually behaves in practice)?

Reply:
We talked about this in chat and resolved it:
- The code will look at many top-level TLVC chunks and early-exit if it finds `FRU0`
- This code path happens if it couldn't find `FRU0`
- While `NoRootChunk` doesn't suggest that to me or @jgallagher, I added a comment on it attempting to explain it.